src/liballoc/boxed.rs

```diff
@@ -511,3 +512,53 @@ impl<'a,A,R> FnOnce<A> for Box<FnBox<A,Output=R>+Send+'a> {
 }
 
 impl<T: ?Sized+Unsize<U>, U: ?Sized> CoerceUnsized<Box<U>> for Box<T> {}
+
+impl<T: Clone> Clone for Box<[T]> {
+    fn clone(&self) -> Self {
+        let mut new = BoxBuilder {
+            data: RawVec::with_capacity(self.len()),
+            len: 0
+        };
+
+        let mut target = new.data.ptr();
+
+        for item in self.iter() {
+            unsafe {
+                ptr::write(target, item.clone());
+                target = target.offset(1);
+            };
+
+            new.len += 1;
+        }
+
+        unsafe { new.into_box() }
+    }
+}
+
+struct BoxBuilder<T> {
+    data: RawVec<T>,
+    len: usize,
+}
+
+impl<T> BoxBuilder<T> {
+    unsafe fn into_box(self) -> Box<[T]> {
+        let raw = ptr::read(&self.data);
+        mem::forget(self);
+        raw.into_box()
+    }
```

Comment:
I just felt philosophically that it was part of BoxBuilder's interface and it was strange for the Clone impl to be messing around like this with the internals of the type when it's custom built for that use case.